core/circuit.go

```diff
@@ -0,0 +1,247 @@
+package core
+
+import (
+	"fmt"
+	"math"
+
+	"github.com/evcc-io/evcc/api"
+	"github.com/evcc-io/evcc/util"
+	"github.com/evcc-io/evcc/util/config"
+)
+
+// the circuit instances to control the load
+type Circuit struct {
+	log    *util.Logger
+	uiChan chan<- util.Param
+
+	maxCurrent    float64           // max allowed current. will not be used when 0
+	maxPower      float64           // max allowed power. will not be used when 0
+	parentCircuit *Circuit          // parent circuit reference, used to determine current limits from hierarchy
+	phaseCurrents api.PhaseCurrents // meter to determine phase current
```

Comment:
Looking at the code below it might be cleaner to have a `meter` and type-check for implemented interfaces.